Add a flag to mailbox table for enable/disable smtp protocol

johnfawcett opened the pull request:
Add a flag to mailbox table for enable/disbale of smtp protocol. the name of the flag is "smtp_active". This enables use of this flag instead of active flag for postfix mysql queries. There is a new configuration parameter $CONF['smtp_active_flag'] = YES or NO, which defaults to NO, providing the existing behaviour. When set to YES the active_smtp field is displayed in the edit mode and can be modified. For simplicity the flag is not shown in the table view, but this could be added. The commit includes update to the language files and documentation. The commit does not include the code to add the field to existing installs, which I presume can be added during the release
 cycle in upgrade.php along the lines of:

_db_add_field('mailbox',  'smtp_active', 'int DEFAULT 1');

DavidGoodwin commented:
Just to be clear - this would allow you to control whether a specific mailbox can send out mail through a postfix server ?

So, if the flag was disabled for my mailbox, I'd be able to receive email from third parties fine, but I wouldn't be able to reply?

Would it be better if this wasn't behind a feature flag, and the smtp_active value just defaults to 1 ?

johnfawcett commented:
Yes, it's a flag that indicates whether a mailbox is allowed to send email or not. In the as-is where this flag does not exist, postfix would typically be configured to use the active flag. Configuring the query in postfix to use the smtp_active flag allows enabling of mail sending separately from mail reading/reception.

I agree the default value of the flag should be  1. The idea behind the configuration option to turn the feature on is for better backwards compatibility. Since in order to use this flag you need to change the postfix query, having the feature turned off will prevent existing installations from seeing the flag in the GUI and being misled into thinking it was being automatically used.

johnfawcett commented on DOCUMENTS/POSTFIX_CONF.txt:
```diff
@@ -140,6 +163,22 @@ hosts = localhost
+password = password
+hosts = localhost
+dbname = postfix
+query = select 'DUNNO' from dual WHERE EXISTS (SELECT a.goto FROM `alias` a join `mailbox` m on a.address=m.username WHERE a.address='%s' and a.active=1 and m.smtp_active=1) UNION select 'REJECT' from dual WHERE NOT EXISTS (SELECT a.goto FROM `alias` a join `mailbox` m on a.address=m.username  WHERE a.address='%s' and a.active=1 and m.smtp_active=1)
```
Yes it is, but looking at it again, a more general way is without "FROM DUAL", i.e. select 'DUNNO' WHERE EXISTS (SELECT a.goto FROM `alias` a join `mailbox` m on a.address=m.username WHERE a.address='%s' and a.active=1 and m.smtp_active=1) UNION select 'REJECT' WHERE NOT EXISTS (SELECT a.goto FROM `alias` a join `mailbox` m on a.address=m.username  WHERE a.address='%s' and a.active=1 and m.smtp_active=1)

DavidGoodwin replied:
Thanks - I did briefly try to write a similar query (without the dual keyword) but I couldn't get the syntax right and then family life took over ....

johnfawcett commented:
I have simplified the documentation with a changed approach. It is possible (and arguably better and certainly simpler) to configure entirely on dovecot if using dovecot sasl in postfix. I reverted the changes on the postfix file and added some comments to the dovecot one.

DavidGoodwin commented on config.inc.php:
```diff
@@ -802,6 +802,8 @@ function x_send_sms_clickatell($to, $message) {
 
 $CONF['version'] = '3.4-dev';
 
+$CONF['smtp_active_flag'] = 'NO';
```
this needs a line or two of docs to explain what the flag is for - perhaps linking to the DOVECOT.txt file etc?

johnfawcett replied:
Thanks for the feedback. I've added a description and link to DOVECOT.txt

DavidGoodwin commented:
It might be worth rebasing your changes onto the upstream master; although I've done that myself locally and the diff seems fine.


I assume for this flag to work, the administrator must have configured the mail server to use a dovecot delivery thingy (listening on port 587 or whatever) instead of using postfix ? Could you re-add the postfix submission 'stuff' - as that might be relevant for others?

(I don't know what advantage there is to using dovecot for  submission/smtp stuff over postfix, I've only ever used postfix)

DavidGoodwin commented:
(Sorry if I sound ungrateful - Thank You for all your work so far)

johnfawcett commented:
I'll take a look at rebasing, but I'm pretty sure the existing commits are ok. Only doubt I have is about the list view modifications, which I tested against the released version. I'll have to see if I can test against the current HEAD since I've merged with other changes.

Regarding submission, I confirm that I'm not using dovecot submission, but postfix submission. With the simplification I did on my configuration, the only change needed to use this flag is on the dovecot side. This is because postfix submission was already configured to use dovecot sasl.

If it is useful I can write something about postfix submission configuration, but it would not be related to the flag as such. Nothing in my postfix configuration now makes use of the new flag, since the task of authentication is handed off to dovecot sasl and I have configured dovecot to give a differentiated response according to the service name (smtp or not).

johnfawcett commented:
I rebased and squashed the previous commits into one 668960ccf6bb9578c9bed6e874a41ddfb30092fb

johnfawcett commented:
Commit 8a314079bd0ff8d86fffaf253d3de77de9fc195b contains the documentation changes to POSTFIX_CONF.txt for configuration of dovecot sasl and submission - these are valid configurations regardless of use of smtp_active flag or existing active flag in the dovecot authentication query

pompfe commented:
> AND (('%s' = 'smtp' AND smtp_active = '1') OR ('%s' <> 'smtp' AND active = '1'))

Just to be sure I understand:

- with smtp_active I control sending only
- with active I control receiving only

Independently of each other.
So, turn off both to not receive and send (disable mailbox). Turn on both to send and receive. And the rest is send-only and receive-only.

Am I correct?

cboltz commented:
@johnfawcett With this PR, you added the following line to POSTFIX_CONF.TXT:

    smtpd_sender_login_maps = proxy:mysql:/etc/postfix/mysql-login_maps_dovecot.cf

Unfortunately you didn't document what query mysql-login_maps_dovecot.cf should contain. Can you please show the content of that file? (here or in a new PR ;-)

DavidGoodwin commented:
see also https://github.com/postfixadmin/postfixadmin/commit/ea11afe80ad5a8770c93e949ac86a2151129b2e1
